The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/26690473007/job/78665911467) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/78665911467) [(plain)](https://github.com/rust-lang/rust/commit/223c0ba48dd39a05b8e25826d6d6caed84916bad/checks/78665911467/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    ╭▸ compiler/rustc_trait_selection/src/error_reporting/infer/note_and_explain.rs:158:77
    │
158 │ …                     // current rust. In fact, when something is refering to this, you
    ╰╴                                                                  ━━━━━━━━
rerun tidy with `--extra-checks=spellcheck --bless` to fix typos
tidy [extra_checks:spellcheck]: checks with external tool 'typos' failed
tidy [extra_checks:spellcheck]: FAIL
yarn install v1.22.22
warning package.json: No license field
warning ../../package.json: License should be a valid SPDX license expression
warning No license field
[1/4] Resolving packages...
---
linting javascript files
Running eslint on rustdoc JS files
info: ES-Check: there were no ES version matching errors!  🎉
typechecking javascript files
tidy: The following check failed: extra_checks:spellcheck
Bootstrap failed while executing `test src/tools/tidy tidyselftest --extra-checks=py,cpp,js,spellcheck`
Command `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools-bin/rust-tidy --root-path=/checkout --cargo-path=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo --output-dir=/checkout/obj/build --concurrency=4 --npm-path=/node/bin/yarn --ci=true --extra-checks=py,cpp,js,spellcheck` failed with exit code 1
Created at: src/bootstrap/src/core/build_steps/tool.rs:1618:23
Executed at: src/bootstrap/src/core/build_steps/test.rs:1417:29

--- BACKTRACE vvv
   0: <bootstrap::utils::exec::DeferredCommand>::finish_process
             at /checkout/src/bootstrap/src/utils/exec.rs:939:17
   1: <bootstrap::utils::exec::DeferredCommand>::wait_for_output::<&bootstrap::utils::exec::ExecutionContext>
             at /checkout/src/bootstrap/src/utils/exec.rs:831:21
   2: <bootstrap::utils::exec::ExecutionContext>::run
             at /checkout/src/bootstrap/src/utils/exec.rs:741:45
   3: <bootstrap::utils::exec::BootstrapCommand>::run::<&bootstrap::core::builder::Builder>
             at /checkout/src/bootstrap/src/utils/exec.rs:339:27
   4: <bootstrap::core::build_steps::test::Tidy as bootstrap::core::builder::Step>::run
             at /checkout/src/bootstrap/src/core/build_steps/test.rs:1417:29
   5: <bootstrap::core::builder::Builder>::ensure::<bootstrap::core::build_steps::test::Tidy>
             at /checkout/src/bootstrap/src/core/builder/mod.rs:1595:36
   6: <bootstrap::core::build_steps::test::Tidy as bootstrap::core::builder::Step>::make_run
             at /checkout/src/bootstrap/src/core/build_steps/test.rs:1339:21
   7: <bootstrap::core::builder::StepDescription>::maybe_run
             at /checkout/src/bootstrap/src/core/builder/mod.rs:476:13
   8: bootstrap::core::builder::cli_paths::match_paths_to_steps_and_run
             at /checkout/src/bootstrap/src/core/builder/cli_paths.rs:232:18
   9: <bootstrap::core::builder::Builder>::run_step_descriptions
             at /checkout/src/bootstrap/src/core/builder/mod.rs:1138:9
  10: <bootstrap::core::builder::Builder>::execute_cli
             at /checkout/src/bootstrap/src/core/builder/mod.rs:1117:14
  11: <bootstrap::Build>::build
             at /checkout/src/bootstrap/src/lib.rs:803:25
  12: bootstrap::main
             at /checkout/src/bootstrap/src/bin/main.rs:130:11
  13: <fn() as core::ops::function::FnOnce<()>>::call_once
             at /rustc/ef0fb8a2563200e322fa4419f09f65a63742038c/library/core/src/ops/function.rs:250:5
  14: std::sys::backtrace::__rust_begin_short_backtrace::<fn(), ()>
             at /rustc/ef0fb8a2563200e322fa4419f09f65a63742038c/library/std/src/sys/backtrace.rs:166:18
  15: std::rt::lang_start::<()>::{closure#0}
             at /rustc/ef0fb8a2563200e322fa4419f09f65a63742038c/library/std/src/rt.rs:206:18
  16: <&dyn core::ops::function::Fn<(), Output = i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe as core::ops::function::FnOnce<()>>::call_once
             at /rustc/ef0fb8a2563200e322fa4419f09f65a63742038c/library/core/src/ops/function.rs:287:21
  17: std::panicking::catch_unwind::do_call::<&dyn core::ops::function::Fn<(), Output = i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe, i32>
             at /rustc/ef0fb8a2563200e322fa4419f09f65a63742038c/library/std/src/panicking.rs:581:40
  18: std::panicking::catch_unwind::<i32, &dyn core::ops::function::Fn<(), Output = i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe>
             at /rustc/ef0fb8a2563200e322fa4419f09f65a63742038c/library/std/src/panicking.rs:544:19
  19: std::panic::catch_unwind::<&dyn core::ops::function::Fn<(), Output = i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe, i32>
             at /rustc/ef0fb8a2563200e322fa4419f09f65a63742038c/library/std/src/panic.rs:359:14
  20: std::rt::lang_start_internal::{closure#0}
             at /rustc/ef0fb8a2563200e322fa4419f09f65a63742038c/library/std/src/rt.rs:175:24
  21: std::panicking::catch_unwind::do_call::<std::rt::lang_start_internal::{closure#0}, isize>
             at /rustc/ef0fb8a2563200e322fa4419f09f65a63742038c/library/std/src/panicking.rs:581:40
---
  28: __libc_start_main
  29: _start


Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:02:41
  local time: Sat May 30 17:42:31 UTC 2026
  network time: Sat, 30 May 2026 17:42:31 GMT
##[error]Process completed with exit code 1.
##[group]Run echo "disk usage:"
```

</details>

--- a/compiler/rustc_trait_selection/src/error_reporting/infer/note_and_explain.rs
+++ b/compiler/rustc_trait_selection/src/error_reporting/infer/note_and_explain.rs
@@ -1,8 +1,7 @@
 use rustc_errors::Applicability::{MachineApplicable, MaybeIncorrect};
 use rustc_errors::{Diag, MultiSpan, pluralize};
-use rustc_hir as hir;
 use rustc_hir::def::DefKind;
-use rustc_hir::find_attr;
+use rustc_hir::{self as hir, LangItem, find_attr};
 use rustc_middle::traits::{ObligationCause, ObligationCauseCode};
 use rustc_middle::ty::error::{ExpectedFound, TypeError};
 use rustc_middle::ty::fast_reject::DeepRejectCtxt;
@@ -140,62 +139,88 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
                             let item_name = tcx.item_name(def_id);
                             let item_args = self.format_generic_args(assoc_args);
 
-                            // Here, we try to see if there's an existing
-                            // trait implementation that matches the one that
-                            // we're suggesting to restrict. If so, find the
-                            // "end", whether it be at the end of the trait
-                            // or the end of the generic arguments.
-                            let mut matching_span = None;
-                            let mut matched_end_of_args = false;
-                            for bound in generics.bounds_for_param(local_id) {
-                                let potential_spans = bound.bounds.iter().find_map(|bound| {
-                                    let bound_trait_path = bound.trait_ref()?.path;
-                                    let def_id = bound_trait_path.res.opt_def_id()?;
-                                    let generic_args = bound_trait_path
-                                        .segments
-                                        .iter()
-                                        .last()
-                                        .map(|path| path.args());
-                                    (def_id == trait_ref.def_id)
-                                        .then_some((bound_trait_path.span, generic_args))
-                                });
+                            if
+                            // if we're referencing an async fn trait's output future
+                            //
+                            // AsyncFnOnce
+                            (tcx.is_lang_item(trait_ref.def_id, LangItem::AsyncFnOnce)
+                                && tcx.is_lang_item(def_id, LangItem::CallOnceFuture))
+                            // AsyncFnMut
+                            ||
+                            (tcx.is_lang_item(trait_ref.def_id, LangItem::AsyncFnMut)
+                                && tcx.is_lang_item(def_id, LangItem::CallRefFuture))
+                            // AsyncFn
+                            ||
+                            (tcx.is_lang_item(trait_ref.def_id, LangItem::AsyncFn)
+                                && tcx.is_lang_item(def_id, LangItem::CallRefFuture))
+                            {
+                                // don't make a suggestion to constrain it, it's not possible in
+                                // current rust. In fact, when something is referring to this, you
+                                // may have just needed to await something.
 
-                                if let Some((end_of_trait, end_of_args)) = potential_spans {
-                                    let args_span = end_of_args.and_then(|args| args.span());
-                                    matched_end_of_args = args_span.is_some();
-                                    matching_span = args_span
-                                        .or_else(|| Some(end_of_trait))
-                                        .map(|span| span.shrink_to_hi());
-                                    break;
-                                }
-                            }
-
-                            if matched_end_of_args {
-                                // Append suggestion to the end of our args
-                                let path = format!(", {item_name}{item_args} = {p}");
-                                note = !suggest_constraining_type_param(
-                                    tcx,
-                                    generics,
-                                    diag,
-                                    &proj.self_ty().to_string(),
-                                    &path,
-                                    None,
-                                    matching_span,
-                                );
+                                diag.help("you may have forgotten to await an async function");
+                                diag.note(format!("it is currently not possible to add bounds constraining the future returned from an async function (`{item_name}`)"));
+                                // don't note, since it's talking about missing bounds. There's
+                                // currently no way to bound the return future
+                                note = false;
                             } else {
-                                // Suggest adding a bound to an existing trait
-                                // or if the trait doesn't exist, add the trait
-                                // and the suggested bounds.
-                                let path = format!("<{item_name}{item_args} = {p}>");
-                                note = !suggest_constraining_type_param(
-                                    tcx,
-                                    generics,
-                                    diag,
-                                    &proj.self_ty().to_string(),
-                                    &path,
-                                    None,
-                                    matching_span,
-                                );
+                                // Here, we try to see if there's an existing
+                                // trait implementation that matches the one that
+                                // we're suggesting to restrict. If so, find the
+                                // "end", whether it be at the end of the trait
+                                // or the end of the generic arguments.
+                                let mut matching_span = None;
+                                let mut matched_end_of_args = false;
+                                for bound in generics.bounds_for_param(local_id) {
+                                    let potential_spans = bound.bounds.iter().find_map(|bound| {
+                                        let bound_trait_path = bound.trait_ref()?.path;
+                                        let def_id = bound_trait_path.res.opt_def_id()?;
+                                        let generic_args = bound_trait_path
+                                            .segments
+                                            .iter()
+                                            .last()
+                                            .map(|path| path.args());
+                                        (def_id == trait_ref.def_id)
+                                            .then_some((bound_trait_path.span, generic_args))
+                                    });
+
+                                    if let Some((end_of_trait, end_of_args)) = potential_spans {
+                                        let args_span = end_of_args.and_then(|args| args.span());
+                                        matched_end_of_args = args_span.is_some();
+                                        matching_span = args_span
+                                            .or_else(|| Some(end_of_trait))
+                                            .map(|span| span.shrink_to_hi());
+                                        break;
+                                    }
+                                }
+
+                                if matched_end_of_args {
+                                    // Append suggestion to the end of our args
+                                    let path = format!(", {item_name}{item_args} = {p}");
+                                    note = !suggest_constraining_type_param(
+                                        tcx,
+                                        generics,
+                                        diag,
+                                        &proj.self_ty().to_string(),
+                                        &path,
+                                        None,
+                                        matching_span,
+                                    );
+                                } else {
+                                    // Suggest adding a bound to an existing trait
+                                    // or if the trait doesn't exist, add the trait
+                                    // and the suggested bounds.
+                                    let path = format!("<{item_name}{item_args} = {p}>");
+                                    note = !suggest_constraining_type_param(
+                                        tcx,
+                                        generics,
+                                        diag,
+                                        &proj.self_ty().to_string(),
+                                        &path,
+                                        None,
+                                        matching_span,
+                                    );
+                                }
                             }
                         }
                         if note {

--- a/tests/ui/async-await/async-closures/wrong-suggestion-result.stderr
+++ b/tests/ui/async-await/async-closures/wrong-suggestion-result.stderr
@@ -25,10 +25,8 @@ LL |     f()
    |
    = note: expected type parameter `O`
              found associated type `<F as AsyncFnOnce<()>>::CallOnceFuture`
-help: consider further restricting this bound
-   |
-LL | async fn once<O, F: AsyncFnOnce(, CallOnceFuture = O) -> O>(f: F) -> O {
-   |                                 ++++++++++++++++++++
+   = help: you may have forgotten to await an async function
+   = note: it is currently not possible to add bounds constraining the future returned from an async function (`CallOnceFuture`)
 
 error[E0277]: the `?` operator can only be used in an async function that returns `Result` or `Option` (or another type that implements `FromResidual`)
   --> $DIR/wrong-suggestion-result.rs:12:21
@@ -57,10 +55,8 @@ LL |     f()
    |
    = note: expected type parameter `O`
              found associated type `<F as AsyncFnMut<()>>::CallRefFuture<'_>`
-help: consider further restricting this bound
-   |
-LL | async fn fnref<O, F: AsyncFn() -> O<CallRefFuture<'_> = O>>(f: F) -> O {
-   |                                    +++++++++++++++++++++++
+   = help: you may have forgotten to await an async function
+   = note: it is currently not possible to add bounds constraining the future returned from an async function (`CallRefFuture`)
 
 error[E0277]: the `?` operator can only be used in an async function that returns `Result` or `Option` (or another type that implements `FromResidual`)
   --> $DIR/wrong-suggestion-result.rs:21:21
@@ -89,10 +85,8 @@ LL |     f()
    |
    = note: expected type parameter `O`
              found associated type `<F as AsyncFnMut<()>>::CallRefFuture<'_>`
-help: consider further restricting this bound
-   |
-LL | async fn fnmut<O, F: AsyncFnMut(, CallRefFuture<'_> = O) -> O>(f: F) -> O {
-   |                                 +++++++++++++++++++++++
+   = help: you may have forgotten to await an async function
+   = note: it is currently not possible to add bounds constraining the future returned from an async function (`CallRefFuture`)
 
 error: aborting due to 6 previous errors
 

